### PR TITLE
Don't push blocks outside world height limits

### LINKED ISSUE #2381: Server crashes at world height with pistons
I was building something that reached the height limit when the entire server crashed. If I had to guess, this crash would be caused by trying to place blocks past the maximum world height, but I could be caused by some other issue related to the max world height. This error can be replicated by simply building a tower up to the height limit.

Here's the output:

```
Error: [ada11208302df6b3|23:35:44] Assertion failed: !"GetBlock(x, y, z) out of bounds!", file <cuberite_dir>/src/Chunk.cpp, line 2625
0   MCServer_debug                      0x00000001079b48ea _Z15PrintStackTracev + 42
1   MCServer_debug                      0x00000001077295e1 _ZNK6cChunk8GetBlockEiii + 145
2   MCServer_debug                      0x000000010774dfdb _ZN9cChunkMap8GetBlockEiii + 1291
3   MCServer_debug                      0x000000010775b4c9 _ZN6cWorld8GetBlockEiii + 73
4   MCServer_debug                      0x0000000107a19299 _ZL32tolua_AllToLua_cWorld_GetBlock00P9lua_State + 329
5   MCServer_debug                      0x0000000107a19c1a _ZL32tolua_AllToLua_cWorld_GetBlock01P9lua_State + 282
6   MCServer_debug                      0x0000000108039718 luaD_precall + 1000
7   MCServer_debug                      0x00000001080546e6 luaV_execute + 7622
8   MCServer_debug                      0x0000000108039f02 luaD_call + 178
9   MCServer_debug                      0x000000010802d9af f_call + 47
10  MCServer_debug                      0x0000000108038d69 luaD_rawrunprotected + 105
11  MCServer_debug                      0x000000010803a35a luaD_pcall + 122
12  MCServer_debug                      0x000000010802d932 lua_pcall + 482
13  MCServer_debug                      0x0000000107a8a29c _ZN9cLuaState12CallFunctionEi + 860
14  MCServer_debug                      0x0000000107af61fc _ZN9cLuaState11PushCallPopIJRbEEEbNS_4cRetEDpOT_ + 44
15  MCServer_debug                      0x0000000107afb3e4 _ZN9cLuaState11PushCallPopIRP7cPlayerJRKNS_4cRetERbEEEbOT_DpOT0_ + 84
16  MCServer_debug                      0x0000000107ab9656 _ZN9cLuaState4CallINS_4cRefEJRP7cPlayerRKNS_4cRetERbEEEbRKT_DpOT0_ + 118
17  MCServer_debug                      0x0000000107ad4234 _ZZN15cManualBindings7ForEachI6cWorld7cPlayerXadL_ZNS1_13ForEachPlayerER13cItemCallbackIS2_EEEEEiP9lua_StateEN12cLuaCallback4ItemEPS2_ + 52
18  MCServer_debug                      0x00000001079467b7 _ZN6cWorld13ForEachPlayerER13cItemCallbackI7cPlayerE + 535
19  MCServer_debug                      0x0000000107ad3fde _ZN15cManualBindings7ForEachI6cWorld7cPlayerXadL_ZNS1_13ForEachPlayerER13cItemCallbackIS2_EEEEEiP9lua_State + 526
20  MCServer_debug                      0x0000000108039718 luaD_precall + 1000
21  MCServer_debug                      0x00000001080546e6 luaV_execute + 7622
22  MCServer_debug                      0x0000000108039f02 luaD_call + 178
23  MCServer_debug                      0x000000010802d9af f_call + 47
24  MCServer_debug                      0x0000000108038d69 luaD_rawrunprotected + 105
25  MCServer_debug                      0x000000010803a35a luaD_pcall + 122
26  MCServer_debug                      0x000000010802d932 lua_pcall + 482
27  MCServer_debug                      0x0000000107a8a29c _ZN9cLuaState12CallFunctionEi + 860
28  MCServer_debug                      0x0000000107af6037 _ZN9cLuaState11PushCallPopEv + 23
29  MCServer_debug                      0x0000000107af98ab _ZN9cLuaState11PushCallPopIRNSt3__16chrono8durationIxNS1_5ratioILl1ELl1000EEEEEJEEEbOT_DpOT0_ + 75
Assertion failed: (0), function GetBlock, file <cuberite_dir>/src/Chunk.cpp, line 2625.
Abort trap: 6
```

### PR DESCRIPTION
**Pull request by mathiascode**
Fixes #2381

### CODE REVIEW

File: src/Blocks/BlockPiston.cpp [hunk 1]
```diff
@@ -160,6 +160,12 @@ bool cBlockPistonHandler::CanPushBlock(
 		return !a_RequirePushable;
 	}
 
+	if ((a_BlockPos.y >= cChunkDef::Height - 1) || (a_BlockPos.y <= 0))
```
Comment (peterbell10) [inline src/Blocks/BlockPiston.cpp]:
```suggestion
	if (!cChunkDef::IsValidHeight(a_BlockPos.y + a_PushDir.y))
```

File: src/Blocks/BlockPiston.cpp [hunk 2]
```diff
@@ -142,9 +142,9 @@ bool cBlockPistonHandler::CanPushBlock(
 	NIBBLETYPE currMeta;
 	a_World.GetBlockTypeMeta(a_BlockPos.x, a_BlockPos.y, a_BlockPos.z, currBlock, currMeta);
 
-	if (currBlock == E_BLOCK_AIR)
+	if ((currBlock == E_BLOCK_AIR) && (a_BlockPos.y <= cChunkDef::Height - 1) && (a_BlockPos.y >= 0))
```
Comment (peterbell10) [inline src/Blocks/BlockPiston.cpp]:
```suggestion
	if ((currBlock == E_BLOCK_AIR) && cChunkDef::IsValidHeight(a_BlockPos.y))
```

File: src/Blocks/BlockPiston.cpp (hunk 1 above)
Comment (peterbell10) [reply src/Blocks/BlockPiston.cpp]: Also, can we handle pushing multiple blocks?

Comment (mathiascode) [reply src/Blocks/BlockPiston.cpp]: Thanks, I missed that function. Yes, that case is handled.

**Comment by mathiascode** [comment]
I got a bit confused by the logic myself. I'll check if your variant works as intended.

**Comment by mathiascode** [comment]
Seems to work.

**Pull request merged**